third_party/nvfuser/csrc/scheduler/registry.cpp

```diff
@@ -2038,10 +2044,272 @@ class PersistentKernelScheduler : public SchedulerEntry {
+          ScheduleHeuristic::Matmul, "No mma op to schedule");
+      return false;
+    }
+
+    // #2
+    // NOTE: this may change in future, with support for C matrix and beta
+    {
+      // TODO: outline to ir_utils namespace
+      const auto& outputs = fusion->outputs();
+      std::vector<Expr*> mmaOutputs;
+
+      for (auto expr : mmaExprs) {
+        const auto& exprOutputs = expr->outputs();
+        if (exprOutputs.size() != 1) {
+          // we are interested only in mma ops with single output val
+          continue;
+        }
+        if (std::find(outputs.cbegin(), outputs.cend(), exprOutputs.front()) !=
+            outputs.cend()) {
+          mmaOutputs.push_back(expr);
+        }
+      }
+
+      mmaExprs = mmaOutputs;
+    }
+
+    if (mmaExprs.empty()) {
+      scheduler_debug_utils::canScheduleRejectReason(
+          ScheduleHeuristic::Matmul, "No output mma op to schedule");
+      return false;
+    }
+
+    // #3
+    {
+      // TODO: outline to ir_utils namespace
+      for (const auto* mmaExpr : mmaExprs) {
+        const auto& mmaExprInputs = mmaExpr->inputs();
+        if (mmaExprInputs.size() != 2) {
+          scheduler_debug_utils::canScheduleRejectReason(
+              ScheduleHeuristic::Matmul,
+              "One of mma op doesn't have 2 inputs, got: " +
+                  std::to_string(mmaExprInputs.size()));
+          return false;
+        }
+
+        // MMA inputs must be TVs
+        if (!mmaExprInputs[0]->isA<TensorView>() ||
+            !mmaExprInputs[1]->isA<TensorView>()) {
+          scheduler_debug_utils::canScheduleRejectReason(
+              ScheduleHeuristic::Matmul, "One of mma op inputs is not a TV");
+          return false;
+        }
+
+        const auto* inA = dynamic_cast<TensorView*>(mmaExprInputs[0]);
+        const auto* inB = dynamic_cast<TensorView*>(mmaExprInputs[1]);
+
+        // The number of IterDomains of MMA inputs must be the same
+        if (inA->nDims() != inB->nDims()) {
+          scheduler_debug_utils::canScheduleRejectReason(
+              ScheduleHeuristic::Matmul,
+              "Mma op inputs don't have the same number of IterDomains, 1st input(" +
+                  std::to_string(inA->nDims()) + "), 2nd input(" +
+                  std::to_string(inB->nDims()) + ")");
+          return false;
+        }
+
+        // The currently supported number of IterDomains perf MMA op input is 3
+        const size_t supportedDims = 3;
+        if (inA->nDims() != supportedDims) {
+          scheduler_debug_utils::canScheduleRejectReason(
+              ScheduleHeuristic::Matmul,
+              "Mma op inputs have unsupported number of IterDomains, got: " +
+                  std::to_string(inA->nDims()) + ", expected " +
+                  std::to_string(supportedDims));
+          return false;
+        }
+
+        // MMA input TVs must be in local memory
+        // NOTE: shouldn't buffers be in shared memory?
```

Comment:
I think what you should be checking here is if the inputs are an input of the fusion. Memory type can be changed for anything that's not input/output to any memory space. Fusion inputs/outputs must be in global memory.

Reply:
Why are they be in Local memory? Aren't we assuming they are in global memory?

Reply:
yeah fusion->exprs() will filter out anything not connected to outputs/inputs so 3 shouldn't be necessary, but might want to check the explicit pattern of input->cast->broadcast->mma, but that's only fully necessary when we allow mma translation in the E2E stack. For now I think these checks are generally good.

But yes, generally you've got the full idea. This check is to validate/limit exactly what the scheduler gets. So when we go to deploy it E2E it should reject anything the scheduler can't handle (in combination with the runtime check).